test_plans/kernel_bundle.asciidoc

```diff
@@ -0,0 +1,139 @@
+** `true` for a kernel that uses 64-bit atomic operations if and only if the selected device has `aspect::atomic64`.
+
+* Define kernels with the following values of the `[[sycl::reqd_work_group_size]]` attribute.
+Verify that `is_compatible` returns true if and only if the work group size is less than or equal to the value reported by `info::device::max_work_group_size` for the selected device.
+
+** `[[sycl::reqd_work_group_size(8)]]`
+** `[[sycl::reqd_work_group_size(16)]]`
+** `[[sycl::reqd_work_group_size(4294967295)]]`
+
+* Define kernels with the following values of the `[[sycl::reqd_sub_group_size]]` attribute.
+Verify that `is_compatible` returns true if and only if the sub group size is not one of the values reported by `info::device::sub_group_sizes` for the selected device.
+
+** `[[sycl::reqd_work_group_size(8)]]`
+** `[[sycl::reqd_work_group_size(16)]]`
+** `[[sycl::reqd_work_group_size(4099)]]`
+
+* for kernels with no optional features used and with `[[sycl::device_has()]]` kernel attribute
+
+** `true` for a kernel decorated with `[[sycl::device_has(aspect::fp16)]]` if and only if the selected device has that aspect.
+** `true` for a kernel decorated with `[[sycl::device_has(aspect::fp64)]]` if and only if the selected device has that aspect.
+** `true` for a kernel decorated with `[[sycl::device_has(aspect::atomic64)]]` if and only if the selected device has that aspect.
+
+* for multiple kernels result is equal to AND operation applied to the multiple calls - each one using single kernel id
+** Check with vector with built-in kernel ids and with the selected device - should return `true`
+** To previous vector add kernel id with large `[[sycl::reqd_work_group_size(4294967295)]]` attribute and with the selected device - should return `false` if value reported by `info::device::max_work_group_size` is less than 4294967295.
+
+
+==== Tests for `template<typename KernelName> bool is_compatible(const device &dev)`
+
+Same tests as previous section except no zero, built-in and multiple kernels cases.
+
+=== `sycl::kernel_bundle` API
+
+==== Deleted default constructor
+
+Check that `std::is_default_constructible_v<sycl::kernel_bundle>` is `false`.
+
+==== Test for `get_backend()`
+
+Call function `kernel_bundle::get_backend()` and check that return type is `sycl::backend`.
+
+==== Tests for `get_devices()`
+
+* Create a 1-element `std::vector<device>` containing the selected device.
+* Create kernel_bundle via `get_kernel_bundle<bundle_state::executable>(const context&, const std::vector<device>&)`.
+* Call `get_devices()`.
+* Check that return type is std::vector<device>.
+* Check that return vector contains only selected device.
+
+==== `has_kernel(const kernel_id &kernelId, const device &dev)`
+
+* Create a kernel bundle in executable state which contains all the application's kernels by calling `get_kernel_bundle<bundle_state::executable>(const context &)`.
```

Comment:
But according to [4.11.12.1. Queries](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:bundles.query) `has_kernel(const kernel_id& kernelId, const device& dev)` 
> Returns: true only if the kernel bundle contains the kernel identified by `kernelId` and if that kernel is compatible with the device `dev`.

So in case of not compatible kernel it returns `false` and `is_compatible` also returns `false`, and if test checks condition "`has_kernel(kernelId, dev)` returns `true` if and only if `is_compatible` is `true`" the test are going to pass when they are both `false`. But with ids of kernels that are compatible with selected device they both would return `true` and test will pass. Am I missing something here?

Reply:
Oh, yes you are right.  I had forgotten that `has_kernel(..., dev)` had that semantic.  The test plan you have is good.

One minor comment is that the statement "which contains all the application's kernels" is not correct, as we discussed in the other comment.

Reply:
Thank you! Fixed.